src/Storage/Database.php

```diff
@@ -0,0 +1,179 @@
+    private $entityManager;
+    /** @var ObjectRepository */
+    private $repository;
+
+    public function __construct(EntityManagerInterface $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+
+    protected function loadEntity($type, $name, $hash): ?string
+    {
+        $data = $this->getRepository()->findOneBy(['type' => $type, 'name' => $name, 'hash' => $hash]);
+        if ($data) {
+            return $data->getValue();
+        }
+
+        return null;
+    }
+
+    /**
+     * @param string $packageName
+     * @param string $hash
+     * @return string|null  Blank if not found.
+     */
+    public function loadPackage(string $packageName, string $hash): ?string
+    {
+        return $this->loadEntity(self::TYPE_PACKAGE, $packageName, $hash);
+    }
+
+    /**
+     * @param string $name
+     * @param string $hash
+     * @return string|null  Blank if not found.
+     */
+    public function loadProvider(string $name, string $hash): ?string
+    {
+        return $this->loadEntity(self::TYPE_PROVIDER, $name, $hash);
+    }
+
+    /**
+     * @return string|null  Blank if not found.
+     */
+    public function loadRoot(): ?string
+    {
+        return $this->loadEntity(self::TYPE_ROOT, '', '');
+    }
+
+    public function loadLatestEntities($type, $names = null)
+    {
+        // we're not interested in the models, only the keyed values, so don't use the repository
+        $qb = new QueryBuilder($this->entityManager);
+        $qb->select('p.name, p.value')
+            ->from(PackageData::class, 'p')
+            ->where('p.type = :type')
+            ->andWhere('p.isLatest = true')
+            ->setParameter('type', $type);
+        if ($names) {
+            $qb->andWhere($qb->expr()->in('p.name', $names));
+        }
+        $data = $qb->getQuery()->getResult(AbstractQuery::HYDRATE_ARRAY);
+        $values = [];
+        foreach ($data as $datum) {
+            $values[$datum['name']] = $datum['value'];
+        }
+
+        return $values;
+    }
+
+    public function loadAllPackages($packageNames)
+    {
+        return $this->loadLatestEntities(self::TYPE_PACKAGE, $packageNames);
+    }
+
+    public function loadAllProviders()
+    {
+        return $this->loadLatestEntities(self::TYPE_PROVIDER);
+    }
+
+    protected function saveEntity(string $type, string $name, string $hash, string $json): bool
+    {
+        // Update or insert as needed.
+        /** @var PackageData[] $data */
+        $data = $this->getRepository()->findBy(['type' => $type, 'name' => $name]);
+        $match = null;
+
+        // ensure there is only one 'latest' package data for this entity
+        foreach ($data as $datum) {
+            if ($datum->getHash() === $hash) {
+                $match = $datum;
+            } elseif ($datum->getIsLatest()) {
+                $datum->setIsLatest(false);
+            } else {
+                $this->entityManager->detach($datum);
```

Comment:
@rasmuswinter did this make a big difference to performance? I'm a bit concerned about depending on the deprecated approach and adding some complexity by having some objects tracked and others not. I would hope Doctrine is efficient enough for this not to make a huge difference (it's not like the issue with flushing where its involvement added DB write operations) but maybe this is overly optimistic.